compiler-core/src/exhaustiveness.rs

```diff
@@ -29,208 +75,629 @@
+                        || name == "String") =>
+            {
+                BranchMode::Infinite
+            }
+
+            Type::Named {
+                module, name, args, ..
+            } if is_prelude_module(module) && name == "List" => BranchMode::List {
+                inner_type: args.first().expect("list has a type argument").clone(),
+            },
 
-    /// A pattern is missing.
-    Failure,
+            Type::Tuple { elems } => BranchMode::Tuple {
+                elems: elems.clone(),
+            },
+
+            Type::Named {
+                module,
+                name,
+                args,
+                inferred_variant,
+                ..
+            } => {
+                let constructors = ConstructorSpecialiser::specialise_constructors(
+                    env.get_constructors_for_type(module, name)
+                        .expect("Custom type variants must exist"),
+                    args.as_slice(),
+                );
 
-    /// Checks if a guard evaluates to true, running the body if it does.
+                let inferred_variant = inferred_variant.map(|i| i as usize);
+                BranchMode::NamedType {
+                    constructors,
+                    inferred_variant,
+                }
+            }
+        }
+    }
+}
+
+/// This is the decision tree that a pattern matching expression gets turned
+/// into: it's a tree-like structure where each path to a root node contains a
+/// series of checks to perform at runtime to understand if a value matches with
+/// a given pattern.
+///
+pub enum Decision {
+    /// This is the final node of the tree, once we get to this one we know we
+    /// have a body to run because a given pattern matched.
     ///
-    /// The arguments are as follows:
+    // todo)) since the tree is not used for code generation, this field is unused.
+    // But it will be useful once we also use this for code gen purposes and not
+    // just for exhaustiveness checking
+    #[allow(dead_code)]
+    Run { body: Body },
+
+    /// We have to make this decision when we run into a branch that also has a
+    /// guard: if it is true we can finally run the body of the branch, stored in
+    /// `if_true`.
+    /// If it is false we might still have to take other decisions and so we might
+    /// have another `DecisionTree` to traverse, stored in `if_false`.
     ///
-    /// 1. The "condition" to evaluate. We just use a dummy value, but in a real
-    ///    compiler this would likely be an AST node of sorts.
-    /// 2. The body to evaluate if the guard matches.
-    /// 3. The sub tree to evaluate when the guard fails.
-    Guard(usize, Body, Box<Decision>),
+    // todo)) since the tree is not used for code generation, the `guard` and
+    // `if_true` fields are unused.
+    // But they will be useful once we also use this for code gen purposes and not
+    // just for exhaustiveness checking
+    #[allow(dead_code)]
```

Comment:
Why is the linter disabled here? This variant is used, as far as I can see.

Reply:
My bad should have been only for the field, not the entire variant!